controller/array_action/array_mediator_svc.py

```diff
@@ -387,12 +386,18 @@ def _delete_volume_by_name(self, volume_name, not_exist_err=True):
     def delete_volume(self, volume_id):
         logger.info("Deleting volume with id : {0}".format(volume_id))
         volume_name = self._get_volume_name_by_wwn(volume_id)
-        fcmap = self._get_fcmap_as_target_if_exists(volume_name)
-        if fcmap:
-            self._stop_and_delete_fcmap(fcmap.id)
-        self._delete_volume_by_name(volume_name)
+        self._delete_object(volume_name)
         logger.info("Finished volume deletion. id : {0}".format(volume_id))
 
+    def _safe_delete_fcmaps_as_source(self, fcmaps_as_source):
+        unfinished_fcmaps = [fcmap for fcmap in fcmaps_as_source
+                             if fcmap.status != FCMAP_STATUS_DONE or fcmap.copy_rate == "0"]
```

Comment:
why is the original `if fcmap.status != FCMAP_STATUS_DONE` not enough?
and even if it's not enough, is the additional check even covered in the unit tests?

Reply:
I just removed the "original"..

Reply:
I covered it in unit tests

Reply:
ok, I see it is covered, but could you just clarify the need for the additional condition `fcmap.copy_rate == "0"` here?

Reply:
check if its not fcmap to snapshot

Reply:
do you mean that before this change, and after snapshot support, there is a bug in volume deletion?
could you elaborate?

Reply:
no, every fcmap as source is assume to be snapshot so in delete_volume only fcmaps as target deleted, now we want to delete fcmap as source too in case its not a snapshot (cause clone exists)

Reply:
but even in case its not a snapshot:

1. why isn't `fcmap.status != FCMAP_STATUS_DONE` enough?
2. why adding `or fcmap.copy_rate == "0"` is enough?

Reply:
there is need for checking snapshot because we cant relay on `idle_or_copied`